src/flatten.jl

```diff
@@ -0,0 +1,181 @@
+                                           c,
+                                           flt_subs,
+                                           subsubcomp_name_to_idx, 
+                                           in_to_idx, 
+                                           out_to_idx)
+                       )
+            end
+        end
+    end
+    return Iterators.flatten(indexed)
+end
+
+function new_names(dest::CompOut, 
+        c, 
+        flt_subs,
+        subsubcomp_name_to_idx, 
+        in_to_idx, 
+        out_to_idx)
+    comp_name = dest.comp_name
+    out_name = dest.out_name
+    subcomp = flt_subs[comp_name]
+    if subcomp isa PrimitiveComponent
+        inp = ()
+    else
+        inp = inputters(subcomp, Output(out_name))
+    end
+    indexed = map(inp) do r
+        if r isa CompOut
+            subcomp_name = subsubcomp_name_to_idx[(dest.comp_name, r.comp_name)]
+            CompOut(subcomp_name, r.out_name)
+        else
+            ()
+        end
+    end
+    return Iterators.flatten(indexed)
+end
+
+function _flatten(c::CompositeComponent)
+    flt_subs_and_mappings = map(_flatten, c.subcomponents)
+    flt_subs = map(first, flt_subs_and_mappings)
+    subcomp_mappings = map(second, flt_subs_and_mappings)
+    # each element of subcomp_mappings is `input_map, output_map` giving the name
+    # maps for a subcomponent
+    idx_to_in = inputs(c) |> keys_deep |> collect
+    idx_to_out = outputs(c) |> keys_deep |> collect
+    in_to_idx = Dict(v => k for (k, v) in enumerate(idx_to_in))
+    out_to_idx = Dict(v => k for (k, v) in enumerate(idx_to_out))
+    idx_to_subsubcomp_name = []
+    for (name, subcomp) in pairs(c.subcomponents)
+        if subcomp isa CompositeComponent 
+            for (innername, _) in pairs(subcomp.subcomponents)
+                push!(idx_to_subsubcomp_name, (name, innername))
+            end
+        else
+            push!(idx_to_subsubcomp_name, (name, name))
+        end
+    end
+    subsubcomp_name_to_idx = Dict(v => k for (k, v) 
+                                  in enumerate(idx_to_subsubcomp_name))
+    new_edges = []
+    for (src, dst) in get_edges(c)
+        srcnames = new_names(src, 
+                             c, 
+                             flt_subs, 
+                             subsubcomp_name_to_idx, 
+                             in_to_idx, 
+                             out_to_idx)
+
+        dstnames = new_names(dst, 
+                             c,
+                             flt_subs, 
+                             subsubcomp_name_to_idx, 
+                             in_to_idx, 
+                             out_to_idx) 
+
+        for (s, d) in Iterators.product(srcnames, dstnames)
+            push!(new_edges, s => d)
+        end
+    end
+    new_comp = CompositeComponent(
+            IndexedValues([inputs(c)[name] for name in idx_to_in]),
+            IndexedValues([outputs(c)[name] for name in idx_to_out]),
+            flt_subs,
```

Comment:
Here, do we need to put `Iterators.flatten(flt_subs)`?

If I'm understanding, `flt_subs` is either a `NTuple{Vector{Component}}` or a `NamedTuple` where the values are `Vector{Component}`, still storing the subsubcomponents grouped together by their original subcomponents.  But we want to end up with the subsubcomponents all flattened out at the same level.